File: easy_equities.py
from requests import Session, get
import urllib.parse
from bs4 import BeautifulSoup
from os import path

class EasyEquities:
	def __init__(self):
		self.baseURL = "https://platform.easyequities.io"
		self.currentAccount = ""
		self.session = Session()

	def login(self, username, password):
		user = urllib.parse.quote(username)
		pw = urllib.parse.quote(password)
		data = (
				f"UserIdentifier={user}&Password={pw}"
				"&ReturnUrl=&OneSignalGameId=&IsUsingNewLayoutSatrixOrEasyEquitiesMobileApp=False"
		)
		self.session.headers["Accept"] = (
			"text/html,application/xhtml+xml,"
			"application/xml;q=0.9,image/webp,*/*;q=0.8"
		)
		self.session.headers["Connection"] = "keep-alive"
		self.session.headers["Connection-Type"] = "application/x-www-form-urlencoded"
		self.session.headers["Content-Type"] = "application/x-www-form-urlencoded"

		response = self.session.post(
			url=self.baseURL + "/Account/SignIn",
			data=data,
			allow_redirects=False
		)
		if (response.status_code != 302):
			print("Login Failed")
			return False

		return True

	def switchAccount(self, accountID):
		if (self.currentAccount == accountID):
			return True

		data = {"trustAccountId": accountID}
		response = self.session.post(
			url=self.baseURL + "/Menu/UpdateCurrency",
			data=data
		)
		if (response.status_code != 200):
			print("Could not select account", accountID)
			return False

		self.currentAccount = accountID
		return True

	def getHoldings(self, accountID):
		if not self.switchAccount(accountID):
			return []

		# Get holdings
		response = self.session.get(url=self.baseURL + "/AccountOverview/GetHoldingsView")
		if (response.status_code != 200):
			print("Could not get holdings")
			return []

		# Parse data
		soup = BeautifulSoup(response.content, "html.parser")
		holding_divs= soup.find_all(attrs={"id": "holding-body-table-positioning"})
		holdings= []
		for holding_div in holding_divs:
			imageURL= holding_div.find(name="img").attrs.get("src")
			image = path.basename(imageURL)
			symbol= image.split(".")[-2]

			purchase_div = holding_div.find(name="div", attrs={"class": "purchase-value-cell"})
			span = purchase_div.find(name="span")
			text = span.text
			purchase_value = float(text[1:].replace(" ", ""))

			current_div = holding_div.find(name="div", attrs={"class": "current-value-cell"})
			span = current_div.find(name="span")
			text = span.text
			current_value = float(text[1:].replace(" ", ""))

			holding = {"symbol": symbol, "purchase_value": purchase_value, "current_value": current_value}
			holdings.append(holding)

		return holdings

	def getBalance(self, accountID):
		if not self.switchAccount(accountID):
			return 0;

		# Get balance
		response = self.session.get(url= self.baseURL + "/AccountOverview")
		if (response.status_code != 200):
			print("Could not get Account overview")
			return 0

		# Parse data
		soup = BeautifulSoup(response.content, "html.parser")
		availableFunds_div = soup.find(attrs={"class": "funds-to-invest"})
		span = availableFunds_div.find(name="span")
		text = span.text
		availableFunds = float(text[1:].replace(" ", ""))
		return availableFunds

	def buy(self, accountID, holding, value):
		if not self.switchAccount(accountID):
			return False

		# Get holding page
		response = self.session.get(url= self.baseURL + "/ValueAllocation/Buy?contractCode=" + holding)
		if (response.status_code != 200):
			print("Could not get Buy page")
			return False

		# Parse data
		soup = BeautifulSoup(response.content, "html.parser")
		form = soup.find(name="form", attrs={"id": "valueAllocationForm"})
		action = form.attrs["action"]
		inputs = form.find_all(name="input")
		postData = {}
		for formInput in inputs:
			name = formInput.attrs["name"]
			value = formInput.attrs["value"]
			if (name == "TradeValue"):
				postData[name] = str(amount)
				continue
			elif (name == "Risk"):
				continue
			postData[name] = value

		# Place order
		response = self.session.post(
			url= self.baseURL+action,
			data=postData
		)
		if (response.status_code != 200):
			print("Could not place order")
			return False

		return True

	def transfer(self, fromAccount, toAccount, fromAmount):
		# Get Transfer page
		response = self.session.get(url= self.baseURL +  "/FundTransfer/Transfer")
		if (response.status_code != 200):
			print("Could not get Transfer page")
			return False

		# Parse data
		soup = BeautifulSoup(response.content, "html.parser")
		form = soup.find(name="form")
		action= form.attrs["action"]
		inputs = form.find_all(name="input")
		postData  = {}
		for formInput in inputs:
			if "name" in formInput.attrs:
				name  = formInput.attrs["name"]
				value = formInput.attrs["value"]
				postData[name] = value

		postData['TransferAmount'] = fromAmount
		postData['SourceTrustAccountId'] = fromAccount
		postData['DestinationTrustAccountId'] = toAccount
		postData['ConfirmAgreement'] = 'true'

		response = self.session.post(
			url= self.baseURL+action,
			data=postData
		)
		if (response.status_code != 200):
			print("Could not transfer funds")
			return False

		return True

File: easy_equities.js
const playwright = require('playwright');
const path = require('path');

const extractNumber = (currencyString) => {
	// Regular expression to find a floating-point number
	// It looks for:
	// - An optional minus sign at the beginning: -?
	// - One or more digits: \d+
	// - An optional decimal part: (\.\d+)?
	const floatRegex = /-?\.?\d+(\s+\d+)?(\.\d+)?/;

	const match = currencyString.match(floatRegex);

	if (match) {
		return parseFloat(match[0].replace(/\s/, ''));
	}

	return 0;
};

class EasyEquities {
	constructor(username, password) {
		this.authentication = { username: username, password: password };
		this.options = {
			headless: true,
			devtools: false,
		};
		this.browser = null;
		this.maxRetries = 3;
	}

	async Initialize() {
		if (this.browser === null) {
			try {
				this.browser = await playwright.firefox.launch(this.options);
				this.context = await this.browser.newContext();
			} catch (error) {
				console.error("Error initializing browser: " + error);
			}
		}
	}

	async Close() {
		if (this.context) {
			await this.context.close();
			this.context = null;
		}

		if (this.browser) {
			await this.browser.close();
			this.browser = null;
		}
	}

	async Login() {
		await this.Initialize();
		const page = await this.context.newPage();

		await page.goto('https://platform.easyequities.io/Account/SignIn');

		let retries = this.maxRetries;
		while (retries > 0) {
			try {
				await page.locator("input#user-identifier-input").fill(this.authentication.username);
				await page.locator("input#Password").fill(this.authentication.password);

				await page.locator("button#SignIn").click();

				try {
					await page.waitForURL(/.+portfolio-overview.+/);
				} catch {
					page.close();
					return false;
				}

				page.close();
				return true;
			} catch (error) {
				retries--;
				console.log("Something went wrong, retrying...");
				if (retries === 0)
					throw new Error(`Failed to get login after ${this.maxRetries} retries: ${error.message}`);
			}
		}
	}

	async SelectAccount(accountID) {
		await this.Initialize();
		const page = await this.context.newPage();

		await page.goto("https://platform.easyequities.io/AccountOverview");

		let retries = this.maxRetries;
		while (retries > 0) {
			try {
				const selectedAccount = await page.locator("h3 > span.bold-heavy").evaluate(element => element.textContent);
				if (selectedAccount.includes(accountID))
					return page;

				await page.locator(`div[data-id='${accountID}']#selector-tab`).click();

				await page.locator(`div[data-id='${accountID}']#trustaccount-slider`).waitFor();
			} catch (error) {
				retries--;
				console.log("Something went wrong, retrying...");
				if (retries === 0)
					throw new Error(`Failed to get holdings for ${accountID} after ${this.maxRetries} retries: ${error.message}`);
			}
		}
	}

	async GetHoldings(accountID) {
		await this.Initialize();
		const page = await this.SelectAccount(accountID);

		let retries = this.maxRetries;
		while (retries > 0) {
			try {
				await page.locator("button#loadHoldings").click();

				const holdings = [];

				await page.locator("div.holdings-group.table-display").waitFor();
				for (const tableDisplay of await page.locator("div.table-display > div#holding-body-table-positioning").all()) {
					// Expand for more information
					tableDisplay.click();

					const holdingValue = await tableDisplay.locator("div.img-stocks-container > img").getAttribute('src');
					const purchaseValue = await tableDisplay.locator("div.purchase-value-cell > span").textContent();
					const currentValue = await tableDisplay.locator("div.current-value-cell > span").textContent();

					await tableDisplay.locator("div.content-box-description").waitFor();
					const detailsTable = await tableDisplay.locator("div.content-box-description > div.row > div > div.content-box > div.row").all();
					const tableValues = await Promise.all(
						detailsTable.map(row => row.locator("div.bold-heavy").textContent())
					);

					const info = {
						symbol: path.basename(holdingValue).split('.')[2],
						purchaseValue: extractNumber(purchaseValue),
						currentValue: extractNumber(currentValue),
						shares: extractNumber(tableValues[0]) + extractNumber(tableValues[1]),
						avgPurchasePrice: extractNumber(tableValues[2]),
					};

					holdings.push(info);
				}

				page.close();
				return holdings;
			} catch (error) {
				retries--;
				console.log("Something went wrong, retrying...");
				if (retries === 0)
					throw new Error(`Failed to get holdings for ${accountID} after ${this.maxRetries} retries: ${error.message}`);
			}
		}
	}

	async GetBalance(accountID) {
		await this.Initialize();
		const page = await this.SelectAccount(accountID);

		let retries = this.maxRetries;
		while (retries > 0) {
			try {
				const availableFunds = await page.locator(`div[data-id='${accountID}'] > div.funds-to-invest > div.bold-heavy`).evaluate(element => element.textContent);

				page.close();
				return extractNumber(availableFunds);
			} catch (error) {
				retries--;
				console.log("Something went wrong, retrying...");
				if (retries === 0)
					throw new Error(`Failed to get balance of ${accountID} after ${this.maxRetries} retries: ${error.message}`);
			}
		}
	}

	async Buy(accountID, holding, amount) {
		await this.Initialize();
		const page = await this.SelectAccount(accountID);

		let retries = this.maxRetries;
		while (retries > 0) {
			try {
				await page.goto("https://platform.easyequities.io/ValueAllocation/Buy?contractCode=" + holding);

				await page.locator("input#js-value-amount").fill(amount.toString());

				const performTradeSelector = "div[data-bind*=performTradeOperation]";
				await page.locator(performTradeSelector).click();

				await page.locator(performTradeSelector).click();

				try {
					await page.waitForURL(/.+BuyInstruction/);
				}
				catch {
					page.close();
					return false;
				}

				page.close();
				return true;
			} catch (error) {
				retries--;
				console.log("Something went wrong, retrying...");
				if (retries === 0)
					throw new Error(`Failed to buy ${holding} after ${this.maxRetries} retries: ${error.message}`);
			}
		}
	}

	async Transfer(fromAccount, toAccount, fromAmount) {
		await this.Initialize();
		const page = await this.context.newPage();

		await page.goto("https://platform.easyequities.io/FundTransfer/Transfer");

		let retries = this.maxRetries;
		while (retries > 0) {
			try {
				await page.locator("select#SourceTrustAccountId").selectOption(fromAccount, { force: true });
				await page.locator("select#DestinationTrustAccountId").selectOption(toAccount, { force: true });

				await page.locator("input#TransferAmount").fill(fromAmount.toString());

				await page.locator("input#Agreements_0__Checked").click();

				await page.locator("button[type='submit']").click();

				try {
					await page.waitForURL(/.+FundTransfer.+successfully.+/);
				}
				catch {
					page.close();
					return false;
				}

				page.close();
				return true;
			} catch (error) {
				retries--;
				console.log("Something went wrong, retrying...");
				if (retries === 0)
					throw new Error(`Failed to get transfer ${fromAmount} from account ${fromAccount} to account ${toAccount} after ${this.maxRetries} retries: ${error.message}`);
			}
		}
	}
};

module.exports = EasyEquities;
